velox/core/PlanNode.cpp

```diff
@@ -429,13 +689,105 @@ void AbstractJoinNode::addDetails(std::stringstream& stream) const {
   }
 }
 
+folly::dynamic AbstractJoinNode::serializeBase() const {
+  auto obj = PlanNode::serialize();
+  obj["joinType"] = joinTypeName(joinType_);
+  obj["leftKeys"] = ISerializable::serialize(leftKeys_);
+  obj["rightKeys"] = ISerializable::serialize(rightKeys_);
+  if (filter_) {
+    obj["filter"] = filter_->serialize();
+  }
+  obj["outputType"] = outputType_->serialize();
+  return obj;
+}
+
+JoinType joinTypeFromName(const std::string& name) {
+  static const std::unordered_map<std::string, JoinType> kJoinTypes = {
+      {"INNER", JoinType::kInner},
+      {"LEFT", JoinType::kLeft},
+      {"RIGHT", JoinType::kRight},
+      {"FULL", JoinType::kFull},
+      {"LEFT SEMI (FILTER)", JoinType::kLeftSemiFilter},
+      {"RIGHT SEMI (FILTER)", JoinType::kRightSemiFilter},
+      {"LEFT SEMI (PROJECT)", JoinType::kLeftSemiProject},
+      {"RIGHT SEMI (PROJECT)", JoinType::kRightSemiProject},
+      {"ANTI", JoinType::kAnti},
+  };
+
+  return kJoinTypes.at(name);
+}
+
 void HashJoinNode::addDetails(std::stringstream& stream) const {
   AbstractJoinNode::addDetails(stream);
   if (nullAware_) {
     stream << ", null aware";
   }
 }
 
+folly::dynamic HashJoinNode::serialize() const {
+  auto obj = serializeBase();
+  obj["nullAware"] = nullAware_;
+  return obj;
+}
+
+// static
+PlanNodePtr HashJoinNode::create(const folly::dynamic& obj, void* context) {
+  auto sources = deserializeSources(obj, context);
+  VELOX_CHECK_EQ(2, sources.size());
+
+  auto nullAware = obj["nullAware"].asBool();
+  auto leftKeys = deserializeFields(obj["leftKeys"], context);
+  auto rightKeys = deserializeFields(obj["rightKeys"], context);
+
+  TypedExprPtr filter;
+  if (obj.count("filter")) {
+    filter = ISerializable::deserialize<ITypedExpr>(obj["filter"]);
+  }
+
+  auto outputType = ISerializable::deserialize<RowType>(obj["outputType"]);
+
+  return std::make_shared<HashJoinNode>(
+      deserializePlanNodeId(obj),
+      joinTypeFromName(obj["joinType"].asString()),
+      nullAware,
+      std::move(leftKeys),
+      std::move(rightKeys),
+      filter,
+      sources[0],
+      sources[1],
+      outputType);
+}
+
+folly::dynamic MergeJoinNode::serialize() const {
+  return serializeBase();
```

Comment:
I would expect this to be what the base implementation of this method would do ialready n case you don't override it?